### ISSUE TITLE
:safe is not safe

### ISSUE THREAD

**Issue by mswrenn** [GaloisInc/cryptol#1044]
Using SBV-backed `:safe` (and `:prove`) on an unsafe term containing `recip` will crash:
```
$ cryptol
version 2.10.0
https://cryptol.net  :? for help

Loading module Cryptol
Cryptol> :safe \(x : Z 7) -> x * (recip x) == 1
Counterexample
cryptol: src/Cryptol/Symbolic/SBV.hs:401:9-37: Non-exhaustive patterns in (vs, [])

$
```
What4 catches the division by 0, though:
```
$ cryptol
version 2.10.0
https://cryptol.net  :? for help

Loading module Cryptol
Cryptol> :set prover = w4-z3
Cryptol> :safe \(x : Z 7) -> x * (recip x) == 1
Counterexample
(\(x : Z 7) -> x * (recip x) == 1) 0 ~> ERROR
(Total Elapsed Time: 0.053s, using "z3")
Cryptol> 
```

**Comment by brianhuffman** [comment]
Here's the place in source code where the failed pattern match is happening:
https://github.com/GaloisInc/cryptol/blob/70442e497f2e74d5877dcca549879dc896cce75d/src/Cryptol/Symbolic/SBV.hs#L420-L427

Basically what's going on here is that in the SBV query, the quantified variables in the Cryptol predicate may have been split up into a larger number of SBV variables with simpler types. SBV returns a counterexample in the form of a list of values with simple types, and then `parseValues` assembles them back together (if necessary) to make values of the possibly-complex Cryptol types.

Usually this parsing process completes with nothing leftover. But it looks like if you use `recip`, the SBV counterexample includes some extra junk. I expect this has to do with the fact that in the SBV backend, `recip` is implemented by generating a new fresh variable for its result value and then asserting something about it.
https://github.com/GaloisInc/cryptol/blob/70442e497f2e74d5877dcca549879dc896cce75d/src/Cryptol/Backend/SBV.hs#L393-L401
The extra junk in the counterexample is probably the values for these variables that were introduced for the result of `recip`. I modified Cryptol to print out the full list of leftover values if the list is non-null; in your example the extra value is the integer `1`. I also found that for each additional occurrence of `recip` you add to the proposition, you get one additional value in that list.

**Comment by robdockins** [comment]
Looks like the same issue as https://github.com/GaloisInc/cryptol/issues/986

I thought that was related to records, but it seems clear that it is actually related to finite field inverses.  I'll see what we can do to fix the counterexample parsing code.

**Comment by robdockins** [comment]
Fixed via #1067
